### PR TITLE
fix: cost of poor quality report time filters not working

### PR DESCRIPTION
**Pull request by aaronmenezes**
#### Issue :
In the Cost of Poor Quality report, Time and date filter had no effect on the report.

#### Fix:
Update report query to fix this incorrect behavior.

### CODE REVIEW

File: erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py [hunk 1]
```diff
@@ -20,30 +19,39 @@ def get_data(report_filters):
...
 		filters = get_filters(report_filters, operations)
+		time_filters = get_timelog_filters(report_filters)
 
-		job_cards = frappe.get_all("Job Card", fields = fields,
-			filters = filters)
-
-		for row in job_cards:
-			row.operating_cost = flt(row.hour_rate) * (flt(row.total_time_in_mins) / 60.0)
-			data.append(row)
+		data = frappe.db.sql("""SELECT {fields} FROM `tabJob Card` tjc {filters}
```
Comment (ankush) [inline erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py]: Translate this to use query builder now 😅 

We are gearing up for Postgres support, so new raw queries won't be allowed from now on. 

https://frappeframework.com/docs/user/en/api/query-builder

Comment (aaronmenezes) [reply erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py]: Okay. I will translate the query to use QB.

**Comment by ankush** [comment]
report not usable right now on `develop` due to date parsing issue, probably introduced via https://github.com/frappe/frappe/pull/13504 

edit: fixed via https://github.com/frappe/frappe/pull/15454

File: erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py [hunk 2]
```diff
@@ -18,31 +17,45 @@ def get_data(report_filters):
...
+					.where(job_card.name.isin(
+						frappe.qb.from_(job_card_time_log)
+						.select(job_card_time_log.parent)
+						.where(job_card_time_log.parent == job_card.name)
+						.where(job_card_time_log.from_time >= report_filters.get('from_date',''))
+						.where(job_card_time_log.to_time <= report_filters.get('to_date',''))
+												))
+				)
+		query = append_filters(report_filters,operations,query,job_card)
+		data = query.run()
```
Comment (ankush) [inline erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py]: the previous function was returning a list of dicts. this is list of tuples. 🤔

File: erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py [hunk 3]
```diff
@@ -18,31 +17,45 @@ def get_data(report_filters):
...
-				filters[field] = report_filters.get(field)
+			if field == 'serial_no':
+				query = query.where(job_Card[field].like('%{}%'.format(report_filters.get(field))))
+			elif field == 'operation':
+				query = query.where(job_Card[field].isin(operations))
 			else:
-				filters[field] = ('like', '% {} %'.format(report_filters.get(field)))
-
-	return filters
+				query = query.where(job_Card[field] ==report_filters.get(field))
```
Comment (ankush) [inline erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py]:
```suggestion
				query = query.where(job_Card[field] == report_filters.get(field))
```

File: erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py [hunk 4]
```diff
@@ -18,31 +17,45 @@ def get_data(report_filters):
...
+						.where(job_card_time_log.to_time <= report_filters.get('to_date',''))
+												))
+				)
+		query = append_filters(report_filters,operations,query,job_card)
+		data = query.run()
 	return data
 
-def get_filters(report_filters, operations):
-	filters = {"docstatus": 1, "operation": ("in", operations), "is_corrective_job_card": 1}
+def append_filters(report_filters, operations,query,job_Card):
```
Comment (ankush) [inline erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py]:
```suggestion
def append_filters(report_filters, operations, query, job_Card):
```

File: erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py [hunk 5]
```diff
@@ -18,31 +17,45 @@ def get_data(report_filters):
...
 	for field in ["name", "work_order", "operation", "workstation", "company", "serial_no", "batch_no", "production_item"]:
```
Comment (ankush) [inline erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py]:
```suggestion
	for field in ("name", "work_order", "operation", "workstation",
					"company", "serial_no", "batch_no", "production_item"):
```

File: erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py [hunk 6]
```diff
@@ -18,31 +17,45 @@ def get_data(report_filters):
...
+					job_card.serial_no, job_card.batch_no, job_card.workstation, job_card.total_time_in_mins, job_card.hour_rate,
+					operatingcost)
+					.where(job_card.docstatus==1)
+					.where(job_card.is_corrective_job_card==1)
+					.where(job_card.name.isin(
+						frappe.qb.from_(job_card_time_log)
+						.select(job_card_time_log.parent)
+						.where(job_card_time_log.parent == job_card.name)
+						.where(job_card_time_log.from_time >= report_filters.get('from_date',''))
+						.where(job_card_time_log.to_time <= report_filters.get('to_date',''))
```
Comment (ankush) [inline erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py]: if comparing an empty string like `x <= ''` nothing will show up I think? 🤔 

These time filters should be conditionally chained like other filters.

File: erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py [hunk 7]
```diff
@@ -18,31 +17,45 @@ def get_data(report_filters):
...
+		job_card = frappe.qb.DocType("Job Card")
+		operatingcost = ((job_card.hour_rate) * (job_card.total_time_in_mins) / 60.0).as_('operating_cost')
+		itemcode = (job_card.production_item).as_('item_code')
+		query = (
+					frappe.qb.from_(job_card)
+					.select(job_card.name, job_card.work_order, itemcode, job_card.item_name, job_card.operation,
+					job_card.serial_no, job_card.batch_no, job_card.workstation, job_card.total_time_in_mins, job_card.hour_rate,
+					operatingcost)
+					.where(job_card.docstatus==1)
+					.where(job_card.is_corrective_job_card==1)
```
Comment (ankush) [inline erpnext/manufacturing/report/cost_of_poor_quality_report/cost_of_poor_quality_report.py]: - This condition should also be added in the subquery for faster filtering. 
- Sub-query should only be required when from/to time is specified.

**Review by ankush — CHANGES_REQUESTED** [review]

**Pull request merged**